featuretools/feature_discovery/feature_discovery.py

```diff
@@ -0,0 +1,413 @@
+
+    Returns:
+        Dict[str, List[Feature]]
+            Hashmap from key to list of features
+
+    Examples:
+        .. code-block:: python
+
+            from featuretools.feature_discovery.feature_discovery import get_features
+            from woodwork.column_schema import ColumnSchema
+
+            f1 = Feature('f1', Double)
+            f2 = Feature('f2', Boolean)
+
+            feature_groups = group_features([f1, f2])
+
+            {
+                "ANY": ["f1", "f2"],
+                "Double": ["f1"],
+                "numeric": ["f1"],
+                "Double,numeric": ["f1"],
+                "Boolean": ["f3"]
+            }
+    """
+    groups = {ANY: []}
+    for feature in features:
+        for key in feature_to_keys(feature=feature):
+            groups.setdefault(key, []).append(feature)
+
+    return groups
+
+
+def get_matching_features(
+    feature_groups: Dict[str, List[Feature]],
+    primitive: Type[PrimitiveBase],
+) -> List[List[Feature]]:
+    """
+    For a given primitive, find all feature sets that can be used to create new feature
+
+    Args:
+        feature_groups (Dict[str, List[Feature]]):
+            Hashmap from Key to List of Features. Key is either: LogicalType name (eg. "Double"), Semantic tag (eg. "index"),
+            or combination (eg. "Double,index").
+        primitive (Type[PrimitiveBase])
+
+    Returns:
+        List[List[Feature]]
+            List of feature sets
+
+    Examples:
+        .. code-block:: python
+
+            from featuretools.feature_discovery.feature_discovery import get_matching_columns
+            from woodwork.column_schema import ColumnSchema
+
+            feature_groups = {
+                "ANY": ["f1", "f2", "f3"],
+                "Double": ["f1", "f2", "f3"],
+                "numeric": ["f1", "f2", "f3"],
+                "Double,numeric": ["f1", "f2", "f3"],
+            }
+
+            feature_sets = get_matching_features(col_groups, AddNumeric)
+
+            [
+                ["f1", "f2"],
+                ["f1", "f3"],
+                ["f2", "f3"]
+            ]
+    """
+    input_sets = primitive.input_types
+    assert input_sets is not None
+    if not isinstance(input_sets[0], list):
+        input_sets = [primitive.input_types]
+
+    commutative = primitive.commutative
+
+    feature_sets = []
+    for input_set in input_sets:
+        assert input_set is not None
+        feature_sets_ = get_features(
+            feature_groups=feature_groups,
+            input_set=input_set,
+            commutative=commutative,
+        )
+
+        feature_sets.extend(feature_sets_)
+
+    return feature_sets
+
+
+def get_primitive_return_type(primitive: Type[PrimitiveBase]) -> ColumnSchema:
+    if primitive.return_type is None:
+        return_type = primitive.input_types[0]
+        if isinstance(return_type, list):
+            return_type = return_type[0]
+    else:
+        return_type = primitive.return_type
+
+    return return_type
```

Comment:
```suggestion
    if primitive.return_type: 
       return primitive.return_type
    return_type = primitive.input_types[0]
    if isinstance(return_type, list):
       return_type = return_type[0]
    return return_type
```
Nit, but IMO this is easier to read. Feel free to disregard